pkg/applicationserver/io/pubsub/pubsub.go

```diff
@@ -17,119 +17,280 @@ package pubsub
+	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/component"
+	"go.thethings.network/lorawan-stack/pkg/errorcontext"
+	"go.thethings.network/lorawan-stack/pkg/errors"
 	"go.thethings.network/lorawan-stack/pkg/log"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
 	"go.thethings.network/lorawan-stack/pkg/unique"
 	"gocloud.dev/pubsub"
 )
 
-type srv struct {
-	ctx           context.Context
-	server        io.Server
-	formatter     formatters.Formatter
-	topics        []*pubsub.Topic
-	subscriptions []*pubsub.Subscription
+// PubSub is an PubSub frontend that exposes ttnpb.ApplicationPubSubRegistryServer.
+type PubSub struct {
+	ttnpb.ApplicationPubSubRegistryServer
+
+	*component.Component
+	ctx      context.Context
+	server   io.Server
+	registry Registry
+
+	integrations      sync.Map
+	integrationErrors sync.Map
 }
 
-// Start starts the pusub frontend.
-func Start(ctx context.Context, server io.Server, formatter formatters.Formatter, pubURLs, subURLs []string) ([]*io.Subscription, error) {
-	ctx = log.NewContextWithField(ctx, "namespace", "applicationserver/io/pubsub")
-	s := &srv{
+// New creates a new pusub frontend.
+func New(c *component.Component, server io.Server, registry Registry) (*PubSub, error) {
+	ctx := log.NewContextWithField(c.FillContext(c.Context()), "namespace", "applicationserver/io/pubsub")
+	ps := &PubSub{
+		Component: c,
 		ctx:       ctx,
 		server:    server,
-		formatter: formatter,
+		registry:  registry,
 	}
+	ps.RegisterTask(ctx, "integrate_all", ps.integrateAll, component.TaskRestartOnFailure)
+	return ps, nil
+}
+
+func (ps *PubSub) integrateAll(ctx context.Context) error {
+	return ps.registry.Range(ctx, []string{"ids"},
+		func(ctx context.Context, _ ttnpb.ApplicationIdentifiers, pb *ttnpb.ApplicationPubSub) bool {
+			ps.startIntegrationTask(ctx, pb.ApplicationPubSubIdentifiers)
+			return true
+		},
+	)
+}
+
+var integrationBackoff = []time.Duration{100 * time.Millisecond, 1 * time.Second, 10 * time.Second}
 
-	subs := []*io.Subscription{}
-	for _, pubURL := range pubURLs {
-		topic, err := pubsub.OpenTopic(ctx, pubURL)
+func (ps *PubSub) startIntegrationTask(ctx context.Context, ids ttnpb.ApplicationPubSubIdentifiers) {
+	ctx = log.NewContextWithFields(ctx, log.Fields(
+		"application_uid", unique.ID(ctx, ids.ApplicationIdentifiers),
+		"pubsub_id", ids.PubSubID,
+	))
+	ps.StartTask(ctx, "integrate", func(ctx context.Context) error {
+		target, err := ps.registry.Get(ctx, ids, ttnpb.ApplicationPubSubFieldPathsNested)
 		if err != nil {
-			return nil, err
+			if !errors.IsNotFound(err) {
+				log.FromContext(ctx).WithError(err).Error("Failed to get integration")
+			}
+			return nil
+		}
+
+		err = ps.integrate(ctx, target)
+		switch {
+		case errors.IsFailedPrecondition(err),
+			errors.IsUnauthenticated(err),
+			errors.IsPermissionDenied(err),
+			errors.IsInvalidArgument(err):
+			log.FromContext(ctx).WithError(err).Warn("Failed to integrate")
```

Comment:
And `Failed to start`; the context is (or should be) clear enough by namespace and log fields